src/pad.rs

```diff
@@ -1,112 +1,213 @@
+            PadMode::Symmetric => (
+                (0..pad).rev().map(|i| i + pad).collect(),
+                (size - pad..size).rev().map(|i| i + pad).collect(),
+            ),
+            PadMode::Wrap => {
+                ((size - pad..size).map(|i| i + pad).collect(), (0..pad).map(|i| i + pad).collect())
+            }
+            _ => panic!("Only Reflect, Symmetric and Wrap have indices"),
+        }
+    }
+}
+
+#[derive(PartialEq)]
+enum PadAction {
+    StopAfterCopy,
+    ByLane,
+    ByIndices,
+    BySides,
+}
+
 /// Pad an image.
 ///
 /// * `data` - A 3D view to the data to pad.
 /// * `pad_width` - Number of values padded to the edges of each axis.
 /// * `mode` - Method that will be used to select the padded values. See the
 ///   [`PadMode`](crate::PadMode) enum for more information.
-pub fn pad<S, A>(data: &ArrayBase<S, Ix3>, pad_width: PadSize, mode: PadMode) -> Array3<A>
+pub fn pad<S, A, D, Sh>(data: &ArrayBase<S, D>, pad: Sh, mode: PadMode<A>) -> Array<A, D>
 where
     S: Data<Elem = A>,
-    A: Zero + Clone + Copy,
+    A: Clone + Copy + FromPrimitive + Num + PartialOrd + std::fmt::Display,
+    D: Dimension + Copy,
+    Sh: ShapeBuilder<Dim = D>,
+    <D as Dimension>::Pattern: NdIndex<D>,
 {
-    let indices = |size: usize, padding: usize| match mode {
-        PadMode::Reflect => reflect_indices(size, padding),
-        PadMode::Symmetric => symmetric_indices(size, padding),
-        PadMode::Wrap => wrap_indices(size, padding),
-    };
-
-    let (width, height, depth) = data.dim();
-    pad_by_indices(
-        data,
-        pad_width,
-        indices(width, pad_width.0),
-        indices(height, pad_width.1),
-        indices(depth, pad_width.2),
-    )
-}
-
-fn reflect_indices(size: usize, padding: usize) -> Vec<u16> {
-    let mut v = Vec::with_capacity(2 * padding + size);
-    let size = size as u16;
-    let padding = padding as u16;
-    v.extend((1..=padding).rev());
-    v.extend(0..size);
-    v.extend((size - padding - 1..size - 1).rev());
-    v
-}
-
-fn symmetric_indices(size: usize, padding: usize) -> Vec<u16> {
-    let mut v = Vec::with_capacity(2 * padding + size);
-    let size = size as u16;
-    let padding = padding as u16;
-    v.extend((0..padding).rev());
-    v.extend(0..size);
-    v.extend((size - padding..size).rev());
-    v
-}
+    let pad = pad.into_shape().raw_dim().clone();
+    let new_dim = data.raw_dim() + pad.clone() + pad.clone();
+    let mut padded = array_like(&data, new_dim, mode.init());
+    let padded_dim = padded.raw_dim();
 
-fn wrap_indices(size: usize, padding: usize) -> Vec<u16> {
-    let mut v = Vec::with_capacity(2 * padding + size);
-    let size = size as u16;
-    let padding = padding as u16;
-    v.extend(size - padding..size);
-    v.extend(0..size);
-    v.extend(0..padding);
-    v
-}
+    // Select portion of padded array that needs to be copied from the original array.
+    let mut orig_portion = padded.view_mut();
+    for d in 0..data.ndim() {
```

Comment:
Could you use https://docs.rs/ndarray/latest/ndarray/struct.ArrayBase.html#method.slice_each_axis_mut
for this.

Reply:
Haha, wow, good find! Yes, `slice_each_axis_mut` replaces the manual loop.